File: kanz/pkg/bus/producer_test.go
package bus_test

import (
	"context"
	"errors"
	"testing"
	"time"

	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/timestamppb"

	envelopepb "github.com/kanz-eng/kanz-schemas-go/envelope/v1"
	"github.com/kanz-eng/kanz/pkg/bus"
)

type captureClient struct {
	sent []bus.Message
}

func (c *captureClient) Publish(_ context.Context, m bus.Message) error {
	c.sent = append(c.sent, m)
	return nil
}
func (c *captureClient) Subscribe(context.Context, string, string, bus.Handler) error {
	return errors.New("not implemented")
}
func (c *captureClient) Close() error { return nil }

func newTestProducer(t *testing.T) (*bus.Producer, *captureClient) {
	t.Helper()
	cc := &captureClient{}
	p, err := bus.NewProducer(cc, bus.ProducerConfig{
		Source:          "test-svc/inst-1",
		ProducerVersion: "test-1.0.0",
	})
	if err != nil {
		t.Fatalf("NewProducer: %v", err)
	}
	return p, cc
}

func factEvent() bus.Event {
	et := time.Date(2026, 5, 15, 12, 0, 0, 0, time.UTC)
	return bus.Event{
		Subject:          "market.equity.trade",
		EventType:        "market.equity.trade",
		EventClass:       envelopepb.EventClass_EVENT_CLASS_FACT,
		SchemaVersion:    1,
		Domain:           "market",
		EventTime:        et,
		PartitionKey:     "AAPL",
		PayloadSchemaRef: "market.v1.MarketDataEvent:1",
		Payload:          timestamppb.New(et),
	}
}

func TestProducerStampsBrokerDedupHeader(t *testing.T) {
	p, cc := newTestProducer(t)
	if err := p.Publish(context.Background(), factEvent()); err != nil {
		t.Fatal(err)
	}
	sent := cc.sent[0]
	env, _, _ := bus.Unframe(sent.Body)
	if env.IdempotencyKey == "" {
		t.Fatal("envelope has empty IdempotencyKey")
	}
	if got := sent.Headers["Nats-Msg-Id"]; got != env.IdempotencyKey {
		t.Errorf("Nats-Msg-Id header=%q want %q (broker-side dedup key)", got, env.IdempotencyKey)
	}
}

func TestProducerStampsFactEvent(t *testing.T) {
	p, cc := newTestProducer(t)
	if err := p.Publish(context.Background(), factEvent()); err != nil {
		t.Fatalf("Publish: %v", err)
	}
	if got := len(cc.sent); got != 1 {
		t.Fatalf("client received %d messages, want 1", got)
	}
	sent := cc.sent[0]
	if sent.Subject != "market.equity.trade" {
		t.Errorf("Subject=%q", sent.Subject)
	}
	if string(sent.Key) != "AAPL" {
		t.Errorf("Key=%q", sent.Key)
	}
	env, payloadBytes, err := bus.Unframe(sent.Body)
	if err != nil {
		t.Fatalf("Unframe: %v", err)
	}
	if env.EventId == "" {
		t.Error("event_id not stamped")
	}
	if env.CorrelationId != env.EventId {
		t.Errorf("CorrelationId=%q want %q (root event)", env.CorrelationId, env.EventId)
	}
	if env.IdempotencyKey != env.EventId {
		t.Errorf("IdempotencyKey=%q want %q (FACT)", env.IdempotencyKey, env.EventId)
	}
	if env.EnvelopeVersion != 1 {
		t.Errorf("EnvelopeVersion=%d want 1", env.EnvelopeVersion)
	}
	if env.Source != "test-svc/inst-1" {
		t.Errorf("Source=%q", env.Source)
	}
	if env.ProducerVersion != "test-1.0.0" {
		t.Errorf("ProducerVersion=%q", env.ProducerVersion)
	}
	if env.ProducerSequence != 1 {
		t.Errorf("ProducerSequence=%d want 1", env.ProducerSequence)
	}
	if env.PublishTime == nil || env.IngestionTime == nil {
		t.Error("PublishTime / IngestionTime not stamped")
	}
	want := timestamppb.New(time.Date(2026, 5, 15, 12, 0, 0, 0, time.UTC))
	got := &timestamppb.Timestamp{}
	if err := proto.Unmarshal(payloadBytes, got); err != nil {
		t.Fatalf("payload unmarshal: %v", err)
	}
	if !proto.Equal(got, want) {
		t.Errorf("payload mismatch: %v vs %v", got, want)
	}
}

func TestProducerPropagatesCausation(t *testing.T) {
	p, cc := newTestProducer(t)
	e := factEvent()
	e.CorrelationID = "root-corr-id"
	e.CausationID = "parent-event-id"
	if err := p.Publish(context.Background(), e); err != nil {
		t.Fatal(err)
	}
	env, _, _ := bus.Unframe(cc.sent[0].Body)
	if env.CorrelationId != "root-corr-id" {
		t.Errorf("CorrelationId=%q want root-corr-id", env.CorrelationId)
	}
	if env.CausationId != "parent-event-id" {
		t.Errorf("CausationId=%q want parent-event-id", env.CausationId)
	}
}

func TestProducerSequenceIncrementsPerPartitionKey(t *testing.T) {
	p, cc := newTestProducer(t)
	for _, pk := range []string{"AAPL", "AAPL", "MSFT", "AAPL"} {
		e := factEvent()
		e.PartitionKey = pk
		if err := p.Publish(context.Background(), e); err != nil {
			t.Fatal(err)
		}
	}
	want := []uint64{1, 2, 1, 3} // AAPL,AAPL,MSFT,AAPL
	for i, m := range cc.sent {
		env, _, _ := bus.Unframe(m.Body)
		if env.ProducerSequence != want[i] {
			t.Errorf("seq[%d]=%d want %d (key=%s)", i, env.ProducerSequence, want[i], env.PartitionKey)
		}
	}
}

func TestProducerSequenceZeroWithoutPartitionKey(t *testing.T) {
	p, cc := newTestProducer(t)
	e := factEvent()
	e.PartitionKey = ""
	if err := p.Publish(context.Background(), e); err != nil {
		t.Fatal(err)
	}
	env, _, _ := bus.Unframe(cc.sent[0].Body)
	if env.ProducerSequence != 0 {
		t.Errorf("ProducerSequence=%d want 0 (no partition_key)", env.ProducerSequence)
	}
}

func TestProducerCommandRequiresIdempotencyKey(t *testing.T) {
	p, _ := newTestProducer(t)
	e := factEvent()
	e.EventClass = envelopepb.EventClass_EVENT_CLASS_COMMAND
	e.EventType = "risk.command.rebalance"
	e.IdempotencyKey = "" // missing on purpose
	if err := p.Publish(context.Background(), e); err == nil {
		t.Fatal("expected error for COMMAND without idempotency_key")
	}
}

func TestProducerCommandPreservesIdempotencyKey(t *testing.T) {
	p, cc := newTestProducer(t)
	e := factEvent()
	e.EventClass = envelopepb.EventClass_EVENT_CLASS_COMMAND
	e.IdempotencyKey = "caller-key-123"
	if err := p.Publish(context.Background(), e); err != nil {
		t.Fatal(err)
	}
	env, _, _ := bus.Unframe(cc.sent[0].Body)
	if env.IdempotencyKey != "caller-key-123" {
		t.Errorf("IdempotencyKey=%q want caller-key-123", env.IdempotencyKey)
	}
	if env.IdempotencyKey == env.EventId {
		t.Error("COMMAND idempotency_key should not equal event_id")
	}
}

func TestProducerRejectsMismatchedIdempotencyKeyForFact(t *testing.T) {
	p, _ := newTestProducer(t)
	e := factEvent()
	e.IdempotencyKey = "not-the-event-id"
	if err := p.Publish(context.Background(), e); err == nil {
		t.Fatal("expected error for FACT with mismatched idempotency_key")
	}
}

func TestProducerRejectsZeroEventTime(t *testing.T) {
	p, _ := newTestProducer(t)
	e := factEvent()
	e.EventTime = time.Time{}
	if err := p.Publish(context.Background(), e); err == nil {
		t.Fatal("expected error for zero EventTime")
	}
}

func TestProducerRejectsNilPayload(t *testing.T) {
	p, _ := newTestProducer(t)
	e := factEvent()
	e.Payload = nil
	if err := p.Publish(context.Background(), e); err == nil {
		t.Fatal("expected error for nil Payload")
	}
}

func TestProducerInheritsLineageFromContext(t *testing.T) {
	p, cc := newTestProducer(t)
	ctx := context.Background()
	ctx = bus.WithCorrelationID(ctx, "corr-from-ctx")
	ctx = bus.WithCausationID(ctx, "parent-evt")
	ctx = bus.WithTraceContext(ctx, "00-ctx-trace")

	if err := p.Publish(ctx, factEvent()); err != nil {
		t.Fatal(err)
	}
	env, _, _ := bus.Unframe(cc.sent[0].Body)
	if env.CorrelationId != "corr-from-ctx" {
		t.Errorf("CorrelationId=%q want corr-from-ctx", env.CorrelationId)
	}
	if env.CausationId != "parent-evt" {
		t.Errorf("CausationId=%q want parent-evt", env.CausationId)
	}
	if env.TraceContext != "00-ctx-trace" {
		t.Errorf("TraceContext=%q want 00-ctx-trace", env.TraceContext)
	}
}

func TestProducerExplicitFieldsBeatContext(t *testing.T) {
	p, cc := newTestProducer(t)
	ctx := bus.WithCorrelationID(context.Background(), "corr-from-ctx")
	ctx = bus.WithCausationID(ctx, "ctx-causation")
	ctx = bus.WithTraceContext(ctx, "ctx-trace")

	e := factEvent()
	e.CorrelationID = "explicit-corr"
	e.CausationID = "explicit-causation"
	e.TraceContext = "explicit-trace"

	if err := p.Publish(ctx, e); err != nil {
		t.Fatal(err)
	}
	env, _, _ := bus.Unframe(cc.sent[0].Body)
	if env.CorrelationId != "explicit-corr" {
		t.Errorf("CorrelationId=%q want explicit-corr (explicit beats ctx)", env.CorrelationId)
	}
	if env.CausationId != "explicit-causation" {
		t.Errorf("CausationId=%q want explicit-causation", env.CausationId)
	}
	if env.TraceContext != "explicit-trace" {
		t.Errorf("TraceContext=%q want explicit-trace", env.TraceContext)
	}
}

func TestProducerRootEventCorrelationDefaultsToEventID(t *testing.T) {
	// No ctx fields, no Event.CorrelationID — root event, correlation = self.
	p, cc := newTestProducer(t)
	if err := p.Publish(context.Background(), factEvent()); err != nil {
		t.Fatal(err)
	}
	env, _, _ := bus.Unframe(cc.sent[0].Body)
	if env.CorrelationId != env.EventId {
		t.Errorf("root CorrelationId=%q want %q", env.CorrelationId, env.EventId)
	}
	if env.CausationId != "" {
		t.Errorf("root CausationId=%q want empty", env.CausationId)
	}
}

func TestNewProducerRequiresSourceAndVersion(t *testing.T) {
	if _, err := bus.NewProducer(&captureClient{}, bus.ProducerConfig{}); err == nil {
		t.Error("expected error for empty config")
	}
	if _, err := bus.NewProducer(&captureClient{}, bus.ProducerConfig{Source: "x"}); err == nil {
		t.Error("expected error for missing ProducerVersion")
	}
	if _, err := bus.NewProducer(nil, bus.ProducerConfig{Source: "x", ProducerVersion: "v"}); err == nil {
		t.Error("expected error for nil client")
	}
}

File: kanz/ts/kanz-bus/test/producer.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";

import { TimestampSchema, timestampFromDate } from "@bufbuild/protobuf/wkt";

import { EventClass } from "@eighred/kanz-schemas/envelope/v1/event_class_pb.js";

import {
  type Client,
  type Event,
  type Handler,
  type Message,
  Producer,
  ProducerConfig,
  unframe,
} from "../src/index.js";

class CaptureClient implements Client {
  readonly sent: Message[] = [];
  async publish(msg: Message): Promise<void> {
    this.sent.push(msg);
  }
  async subscribe(_s: string, _g: string, _h: Handler): Promise<void> {
    throw new Error("not implemented");
  }
  async close(): Promise<void> {}
}

function makeProducer(): { producer: Producer; client: CaptureClient } {
  const client = new CaptureClient();
  const producer = new Producer(client, {
    source: "test-svc/inst-1",
    producerVersion: "test-1.0.0",
  });
  return { producer, client };
}

function factEvent(): Event {
  const et = new Date("2026-05-15T12:00:00Z");
  return {
    subject: "market.equity.trade",
    eventType: "market.equity.trade",
    eventClass: EventClass.FACT,
    schemaVersion: 1,
    domain: "market",
    eventTime: et,
    partitionKey: "AAPL",
    payloadSchemaRef: "market.v1.MarketDataEvent:1",
    payload: { schema: TimestampSchema, message: timestampFromDate(et) },
  };
}

test("producer stamps FACT event", async () => {
  const { producer, client } = makeProducer();
  await producer.publish(factEvent());
  assert.equal(client.sent.length, 1);
  const sent = client.sent[0]!;
  assert.equal(sent.subject, "market.equity.trade");
  assert.deepEqual(sent.key, new TextEncoder().encode("AAPL"));

  const { envelope: env } = unframe(sent.body);
  assert.ok(env.eventId);
  assert.equal(env.correlationId, env.eventId); // root event
  assert.equal(env.idempotencyKey, env.eventId); // FACT
  assert.equal(env.envelopeVersion, 1);
  assert.equal(env.source, "test-svc/inst-1");
  assert.equal(env.producerVersion, "test-1.0.0");
  assert.equal(env.producerSequence, 1n);
  assert.ok(env.publishTime);
  assert.ok(env.ingestionTime);
});

test("producer propagates explicit causation", async () => {
  const { producer, client } = makeProducer();
  const e = factEvent();
  e.correlationId = "root-corr-id";
  e.causationId = "parent-event-id";
  await producer.publish(e);
  const { envelope: env } = unframe(client.sent[0]!.body);
  assert.equal(env.correlationId, "root-corr-id");
  assert.equal(env.causationId, "parent-event-id");
});

test("producer sequence increments per partition_key", async () => {
  const { producer, client } = makeProducer();
  for (const pk of ["AAPL", "AAPL", "MSFT", "AAPL"]) {
    const e = factEvent();
    e.partitionKey = pk;
    await producer.publish(e);
  }
  const seqs = client.sent.map((m) => unframe(m.body).envelope.producerSequence);
  assert.deepEqual(seqs, [1n, 2n, 1n, 3n]);
});

test("producer sequence is 0 without partition_key", async () => {
  const { producer, client } = makeProducer();
  const e = factEvent();
  e.partitionKey = "";
  await producer.publish(e);
  const { envelope: env } = unframe(client.sent[0]!.body);
  assert.equal(env.producerSequence, 0n);
});

test("COMMAND requires idempotency_key", async () => {
  const { producer } = makeProducer();
  const e = factEvent();
  e.eventClass = EventClass.COMMAND;
  e.eventType = "risk.command.rebalance";
  e.idempotencyKey = "";
  await assert.rejects(producer.publish(e), /idempotency_key/);
});

test("COMMAND preserves caller idempotency_key", async () => {
  const { producer, client } = makeProducer();
  const e = factEvent();
  e.eventClass = EventClass.COMMAND;
  e.idempotencyKey = "caller-key-123";
  await producer.publish(e);
  const { envelope: env } = unframe(client.sent[0]!.body);
  assert.equal(env.idempotencyKey, "caller-key-123");
  assert.notEqual(env.idempotencyKey, env.eventId);
});

test("non-COMMAND rejects mismatched idempotency_key", async () => {
  const { producer } = makeProducer();
  const e = factEvent();
  e.idempotencyKey = "not-the-event-id";
  await assert.rejects(producer.publish(e), /idempotency_key/);
});

test("producer rejects missing eventTime", async () => {
  const { producer } = makeProducer();
  const e = factEvent();
  (e as unknown as { eventTime: undefined }).eventTime = undefined;
  await assert.rejects(producer.publish(e), /eventTime/);
});

test("producer rejects missing payload", async () => {
  const { producer } = makeProducer();
  const e = factEvent();
  (e as unknown as { payload: undefined }).payload = undefined;
  await assert.rejects(producer.publish(e), /payload/);
});

test("Producer constructor rejects missing config / client", () => {
  assert.throws(
    () => new Producer(new CaptureClient(), { source: "", producerVersion: "" } as ProducerConfig),
    /source/,
  );
  assert.throws(
    () => new Producer(new CaptureClient(), { source: "x", producerVersion: "" } as ProducerConfig),
    /producerVersion/,
  );
  assert.throws(
    () =>
      new Producer(null as unknown as Client, {
        source: "x",
        producerVersion: "v",
      }),
    /client/,
  );
});

test("producer stamps broker-dedup header", async () => {
  const { producer, client } = makeProducer();
  await producer.publish(factEvent());
  const sent = client.sent[0]!;
  const { envelope: env } = unframe(sent.body);
  assert.ok(env.idempotencyKey);
  assert.equal(sent.headers?.["Nats-Msg-Id"], env.idempotencyKey);
});

test("root event correlation_id defaults to event_id", async () => {
  const { producer, client } = makeProducer();
  await producer.publish(factEvent());
  const { envelope: env } = unframe(client.sent[0]!.body);
  assert.equal(env.correlationId, env.eventId);
  assert.equal(env.causationId, "");
});
